### ISSUE TITLE
[Bug][runtime-env] Failed to fetch URI gcs://_ray_pkg_23ac9932199ad6d4.zip from GCS.

### ISSUE THREAD

**Issue by ankur6ue** [ray-project/ray#23423]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### Issue Severity

Medium: It contributes to significant difficulty to complete my task but I work arounds and get it resolved.

### What happened + What you expected to happen

While connecting to a Ray cluster running on Kubernetes using ray client, I occasionally (~40% of the time) get an error: Eg:

error message: Failed to fetch URI gcs://_ray_pkg_23ac9932199ad6d4.zip from GCS.

The cluster is unable to download the zip of my working_dir, as specified in the runtime_env argument to ray.init

If I make a trivial change to the code in the working directory, ray client creates a new package, which usually works. 

### Versions / Dependencies

happens on ray 1.11

### Reproduction script

  ray.init(address="address of loadbalancer exposing the ray service on kubernetes:10001",
             runtime_env={"working_dir": working_dir,
                          "excludes":["/models/*", ".cache/", "*.pkl", "*.flac"],
                          "env_vars":config})

### Anything else

Happens about 40% of the time

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

**Comment by scv119** [comment]
Looks a runtime env related issue. @architkulkarni have you seen similar issues and any insight how to debug this?
@ankur6ue I wonder if you have driver logs showing why exactly this fetch failed?

**Comment by ankur6ue** [comment]
Where should I look for these driver logs?

**Comment by architkulkarni** [comment]
Hi @ankur6ue, thanks for reporting and sorry you're running into this.  This sounds like it might be a bug with the way we're garbage collecting the working_dir resources on each node when the client connection ends.  There are logs on the head node in `dashboard_agent.log` and `runtime_env_setup_*.log` but I'm not sure they will have more details.  

To help debugging this, can you set the environment variable `RAY_runtime_env_skip_local_gc` to `1` on each node in your cluster before starting Ray and see if that affects things?  Alternatively, is it possible for you to try out the [nightly wheels](https://docs.ray.io/en/latest/ray-overview/installation.html#daily-releases-nightlies) on your local machine and on the cluster  and see if the problem persists?

**Comment by ankur6ue** [comment]
I tried setting RAY_runtime_env_skip_local_gc to 1 but still getting random GCS errors:
error message: Failed to fetch URI gcs://_ray_pkg_bf5c39d1ac02aa14.zip from GCS.

They seem to be somewhat less frequent, but still occurring 25-30% of the time

**Comment by architkulkarni** [comment]
Thanks for testing it out!  Very interesting to see these messages even with the GC turned off.  We'll try to reproduce this on our end.

**Comment by shrekris-anyscale** [comment]
I believe `RAY_runtime_env_skip_local_gc` may have no effect. PR [#20789](https://github.com/ray-project/ray/pull/20789) seems to have removed the logic that skips garbage collection using that environment variable ([relevant diff](https://github.com/ray-project/ray/commit/32cf19a8815348894319fa2ea8f72c97861768f3#diff-d2f22b8f1bf5f9be47dacae8b467a72ee94629df12ffcc18b13447192ff3dbcfL302-L308)).

I searched the repo, but the only code using `"runtime_env_skip_local_gc"` was [where its default `false` value was set](https://github.com/ray-project/ray/blob/781c46ae448708f90f21236a065b4d68089876eb/src/ray/common/ray_config_def.h#L551). This may be a wider issue since the [documentation](https://github.com/ray-project/ray/blob/990b0ec93420973939c0d68284a774cfc6d4798f/doc/source/ray-core/handling-dependencies.rst) and some tests ([file 1](https://github.com/ray-project/ray/blob/cb80518a8090e26f809a97de4a15416fab5f0bc0/python/ray/tests/test_runtime_env_conda_and_pip_3.py#L150-L159), [file 2](https://github.com/ray-project/ray/blob/8707eb628843a6ab02e6452f487bb7f7df8a5e8f/python/ray/tests/test_runtime_env_working_dir_2.py#L545-L554)) still use the variable.

**Comment by blublinsky** [comment]
Is there any resolution, that will fix this?

**Comment by blublinsky** [comment]
Doing more experiments. It looks like:
1. CLI is working most of the time, while Python code seems to be broken.
2. When using CLI with Kubernetes cluster a strange thing is that it works with port-forwarding, but fails to connect when using ingress, while Python works with ingress fine

**Comment by blublinsky** [comment]
And the last point is that it used to work fine with 1.10

**Comment by architkulkarni** [comment]
Hi @blublinsky, thanks for the info and sorry you're running into this. Can you share any more details about your setup? Also, I'm not sure what you mean by CLI vs Python--which CLI command and which python code are you running?

**Comment by blublinsky** [comment]
I am using latest version of APIs:
````
pip3 show ray
Name: ray
Version: 1.12.0
Summary: Ray provides a simple, universal API for building distributed applications.
Home-page: https://github.com/ray-project/ray
Author: Ray Team
Author-email: ray-dev@googlegroups.com
License: Apache 2.0
Location: /usr/local/lib/python3.9/site-packages
Requires: aiosignal, attrs, click, filelock, frozenlist, grpcio, jsonschema, msgpack, numpy, protobuf, pyyaml, requests, virtualenv
````

This command mostly work:

````
ray job submit --runtime-env-json='{"working_dir": "/Users/boris/Projects/Platform-Infrastructure/middleware/ray/test/job_client", "pip": ["requests==2.26.0", "qiskit==0.34.2"], "env_vars": {"MY_VARIABLE": "foo"}}' -- python script_with_parameters.py --kwargs iterations=7
````

But this code does (which does the same thing) not:

````
from ray.dashboard.modules.job.sdk import JobSubmissionClient
from ray.dashboard.modules.job.common import JobStatus, JobStatusInfo

import time

client = JobSubmissionClient("http://127.0.0.1:8265")
#client = JobSubmissionClient("http://raycluster-ray.borisdata-cluster-2a0beb393d3242574412e5315d3d4662-0000.us-south.containers.appdomain.cloud")

job_id = client.submit_job(
    # Entrypoint shell command to execute
    entrypoint="python script_with_parameters.py --kwargs iterations=7",
    # Working dir
    runtime_env={
        "working_dir": "/Users/boris/Projects/Platform-Infrastructure/middleware/ray/test/job_client",
        "pip": ["requests==2.26.0", "qiskit==0.34.2"],
        "env_vars": {"MY_VARIABLE": "foo"}
    }
)

print(f"Submitted job with ID : {job_id}")

while True:
    status_info = client.get_job_status(job_id)
    status = status_info.status
    print(f"status: {status}")
    if status in {JobStatus.SUCCEEDED, JobStatus.STOPPED, JobStatus.FAILED}:
        break
    time.sleep(5)

logs = client.get_job_logs(job_id)
print(f"logs: {logs}")
````

**Comment by blublinsky** [comment]
The actual code that I am invoking is:
````
import ray
import os
import requests
import qiskit
import argparse

class ParseKwargs(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        setattr(namespace, self.dest, dict())
        for value in values:
            key, value = value.split('=')
            getattr(namespace, self.dest)[key] = value

parser = argparse.ArgumentParser()
parser.add_argument('-k', '--kwargs', nargs='*', action=ParseKwargs)
args = parser.parse_args()

numberOfIterations = int(args.kwargs["iterations"])
print(f"Requested number of iterations is: {numberOfIterations}")

print(f'Envarionment variable MY_VARIABLE has a value of {os.getenv("MY_VARIABLE")}')

ray.init()

@ray.remote
class Counter:
    def __init__(self):
        self.counter = 0

    def inc(self):
        self.counter += 1

    def get_counter(self):
        return self.counter

counter = Counter.remote()

for _ in range(numberOfIterations):
    ray.get(counter.inc.remote())
    print(ray.get(counter.get_counter.remote()))

print("Requests", requests.__version__)
print("Qiskit", qiskit.__version__)

````

**Comment by blublinsky** [comment]
@architkulkarni, after experimenting more it turns out that the problem was my environment, that got messed up. It seems to work fine now

**Comment by architkulkarni** [comment]
@blublinsky No problem, I'd still be curious to hear details about the root cause.  It might help other users too.  And we can try to surface a better error message than "Failed to fetch URI gcs://..."

**Comment by blublinsky** [comment]
I messed python installation on my machine. After I fixed it, it worked. The only caveat that I have noticed in the latest version is that now, if URL is provided without the port, you attach port 8265 which breaks default ingress. Now I need to explicitly add port 80

**Comment by architkulkarni** [comment]
@ankur6ue do you know if your problem was resolved with Ray 1.12?  If not, could you provide some more details about how your Ray Kubernetes cluster is set up?  We've tried to reproduce this using Ray 1.11 on a multiple node cluster, but weren't successful.

**Comment by peterhaddad3121** [comment]
@architkulkarni I am am experiencing problems after uploading to version 1.12.0 of Ray from 1.11.0.

I am using the job SDK and receive errors such as: 

```
OSError: Failed to fetch URI gcs://_ray_pkg_a7fff72d976d2f6a.zip from GCS.
2022-04-25 03:45:31,411 ERROR runtime_env_agent.py:287 -- Failed to create runtime env {"uris": {"workingDirUri": "gcs://_ray_pkg_a7fff72d976d2f6a.zip"}, "workingDir": "gcs://_ray_pkg_a7fff72d976d2f6a.zip"}.
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 281, in CreateRuntimeEnv
    runtime_env_setup_task, timeout=setup_timeout_seconds
  File "/home/ray/anaconda3/lib/python3.7/asyncio/tasks.py", line 442, in wait_for
    return fut.result()
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 151, in _setup_runtime_env
    uri, runtime_env, context, logger=per_job_logger
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/runtime_env/working_dir.py", line 150, in create
    return await loop.run_in_executor(None, _create)
  File "/home/ray/anaconda3/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/runtime_env/working_dir.py", line 145, in _create
    uri, self._resources_dir, logger=logger
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/runtime_env/packaging.py", line 480, in download_and_unpack_package
    raise IOError(f"Failed to fetch URI {pkg_uri} from GCS.")
OSError: Failed to fetch URI gcs://_ray_pkg_a7fff72d976d2f6a.zip from GCS.
2022-04-25 03:45:32,412 ERROR runtime_env_agent.py:298 -- Runtime env creation failed for 3 times, don't retry any more.
```

Also, I am using the Job SDK and when printing the status I receive this error (which seems very misleading from the client side).

```
TypeError: __init__() got an unexpected keyword argument 'entrypoint'
```

Any help would be appreciated. I am using the exact RL script from 1.11.0, so i'm uncertain if my solution will be the same as @blublinsky

**Comment by architkulkarni** [comment]
Yes @peterhaddad3121 it would be great if you could make a separate issue.  In the issue would you also mind adding more details about the port concern and why it's limiting; i.e. is it an inconvenience or a blocker?  What does it mean for port 8265 to break default ingress?  In any case we can definitely try to improve the experience.

**Comment by han14466** [comment]
I found the same error in 1.10.0
I attached my dashboard_agent.log below. Maybe first time, execute ray.init and quit . After several seconds, execute  ray.init again. Ray found same zip already in /tmp/ray/path, at sametime runtime_env_agent async  DeleteURIs. So it throws these errors. 
2022-04-28 16:06:05,744 INFO runtime_env_agent.py:210 -- Successfully created runtime env: {"pyModules": ["gcs://_ray_pkg_d474af55b055c082.zip", "gcs://_ray_pkg_cd17b9c796895104.zip"], "uris": {"pyModulesUris": ["gcs://_ray_pkg_d474af55b055c082.zip", "gcs://_ray_pkg_cd17b9c796895104.zip"]}}, the context: {"command_prefix": [], "env_vars": {"PYTHONPATH": "/tmp/ray/session_2022-04-26_14-42-39_526087_8514/runtime_resources/py_modules_files/_ray_pkg_d474af55b055c082:/tmp/ray/session_2022-04-26_14-42-39_526087_8514/runtime_resources/py_modules_files/_ray_pkg_cd17b9c796895104"}, "py_executable": "/usr/local/bin/python3.6", "resources_dir": null, "container": {}}
2022-04-28 16:06:08,561 INFO runtime_env_agent.py:162 -- Runtime env already created successfully. Env: {"pyModules": ["gcs://_ray_pkg_d474af55b055c082.zip", "gcs://_ray_pkg_49eb4731375affe2.zip"], "uris": {"pyModulesUris": ["gcs://_ray_pkg_d474af55b055c082.zip", "gcs://_ray_pkg_49eb4731375affe2.zip"]}}, context: {"command_prefix": [], "env_vars": {"PYTHONPATH": "/tmp/ray/session_2022-04-26_14-42-39_526087_8514/runtime_resources/py_modules_files/_ray_pkg_d474af55b055c082:/tmp/ray/session_2022-04-26_14-42-39_526087_8514/runtime_resources/py_modules_files/_ray_pkg_49eb4731375affe2"}, "py_executable": "/usr/local/bin/python3.6", "resources_dir": null, "container": {}}
2022-04-28 16:06:08,562 INFO runtime_env_agent.py:216 -- Got request to delete URIs: ['py_modules|gcs://_ray_pkg_d474af55b055c082.zip'].
2022-04-28 16:06:08,581 INFO runtime_env_agent.py:216 -- Got request to delete URIs: ['py_modules|gcs://_ray_pkg_49eb4731375affe2.zip'].
2022-04-28 16:06:10,349 INFO runtime_env_agent.py:179 -- Creating runtime env: {"pyModules": ["gcs://_ray_pkg_d474af55b055c082.zip", "gcs://_ray_pkg_cd17b9c796895104.zip"], "uris": {"pyModulesUris": ["gcs://_ray_pkg_d474af55b055c082.zip", "gcs://_ray_pkg_cd17b9c796895104.zip"]}}
2022-04-28 16:06:10,356 ERROR runtime_env_agent.py:189 -- Runtime env creation failed.
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 186, in CreateRuntimeEnv
    request.serialized_allocated_resource_instances)
  File "/usr/local/lib/python3.6/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 148, in _setup_runtime_env
    return await loop.run_in_executor(None, run_setup_with_logger)
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.6/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 110, in run_setup_with_logger
    runtime_env, context, logger=per_job_logger)
  File "/usr/local/lib/python3.6/site-packages/ray/_private/runtime_env/py_modules.py", line 116, in setup
    uri, self._resources_dir, logger=logger)
  File "/usr/local/lib/python3.6/site-packages/ray/_private/runtime_env/packaging.py", line 435, in download_and_unpack_package
    raise IOError(f"Failed to fetch URI {pkg_uri} from GCS.")
OSError: Failed to fetch URI gcs://_ray_pkg_d474af55b055c082.zip from GCS.
2022-04-28 16:06:11,361 ERROR runtime_env_agent.py:189 -- Runtime env creation failed.
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 186, in CreateRuntimeEnv
    request.serialized_allocated_resource_instances)
  File "/usr/local/lib/python3.6/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 148, in _setup_runtime_env
    return await loop.run_in_executor(None, run_setup_with_logger)
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.6/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 110, in run_setup_with_logger
    runtime_env, context, logger=per_job_logger)
  File "/usr/local/lib/python3.6/site-packages/ray/_private/runtime_env/py_modules.py", line 116, in setup
    uri, self._resources_dir, logger=logger)
  File "/usr/local/lib/python3.6/site-packages/ray/_private/runtime_env/packaging.py", line 435, in download_and_unpack_package
    raise IOError(f"Failed to fetch URI {pkg_uri} from GCS.")
OSError: Failed to fetch URI gcs://_ray_pkg_d474af55b055c082.zip from GCS.
2022-04-28 16:06:12,456 ERROR runtime_env_agent.py:189 -- Runtime env creation failed.
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 186, in CreateRuntimeEnv
    request.serialized_allocated_resource_instances)
  File "/usr/local/lib/python3.6/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 148, in _setup_runtime_env
    return await loop.run_in_executor(None, run_setup_with_logger)
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.6/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 110, in run_setup_with_logger
    runtime_env, context, logger=per_job_logger)
  File "/usr/local/lib/python3.6/site-packages/ray/_private/runtime_env/py_modules.py", line 116, in setup
    uri, self._resources_dir, logger=logger)
  File "/usr/local/lib/python3.6/site-packages/ray/_private/runtime_env/packaging.py", line 435, in download_and_unpack_package
    raise IOError(f"Failed to fetch URI {pkg_uri} from GCS.")
OSError: Failed to fetch URI gcs://_ray_pkg_d474af55b055c082.zip from GCS.
2022-04-28 16:06:13,477 ERROR runtime_env_agent.py:198 -- Runtime env creation failed for 3 times, don't retry any more.
2022-04-28 16:06:13,478 INFO runtime_env_agent.py:169 -- Runtime env already failed. Env: {"pyModules": ["gcs://_ray_pkg_d474af55b055c082.zip", "gcs://_ray_pkg_cd17b9c796895104.zip"], "uris": {"pyModulesUris": ["gcs://_ray_pkg_d474af55b055c082.zip", "gcs://_ray_pkg_cd17b9c796895104.zip"]}}, err: Failed to fetch URI gcs://_ray_pkg_d474af55b055c082.zip from GCS.

**Comment by christianvogellive** [comment]
is there any quick workaround for this issue currently? or do we need to wait for the PR? 
kind of stuck because of this issue and already tried out different versions.

**Comment by Nintorac** [comment]
I too have experienced this, I am using ray 1.12.0. Not sure how related our root causes are however.

I got it after deleting the nodes using AWS console, the nodes I deleted were the ones hosting my two worker pods.

I was able to restore the functionality by deleting the head pod.

My traceback is as follows

```bash
Traceback (most recent call last):
  File "tune_runner.py", line 49, in main
    ray.get(remote_tune.remote(config))
  File "/home/oliver/miniconda3/envs/nvflare/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 104, in wrapper
    return getattr(ray, func.__name__)(*args, **kwargs)
  File "/home/oliver/miniconda3/envs/nvflare/lib/python3.8/site-packages/ray/util/client/api.py", line 43, in get
    return self.worker.get(vals, timeout=timeout)
  File "/home/oliver/miniconda3/envs/nvflare/lib/python3.8/site-packages/ray/util/client/worker.py", line 433, in get
    res = self._get(to_get, op_timeout)
  File "/home/oliver/miniconda3/envs/nvflare/lib/python3.8/site-packages/ray/util/client/worker.py", line 461, in _get
    raise err
ray.exceptions.RuntimeEnvSetupError: Failed to setup runtime environment.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 280, in CreateRuntimeEnv
    runtime_env_context = await asyncio.wait_for(
  File "/opt/conda/lib/python3.8/asyncio/tasks.py", line 494, in wait_for
    return fut.result()
  File "/opt/conda/lib/python3.8/site-packages/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 150, in _setup_runtime_env
    size_bytes = await manager.create(
  File "/opt/conda/lib/python3.8/site-packages/ray/_private/runtime_env/working_dir.py", line 150, in create
    return await loop.run_in_executor(None, _create)
  File "/opt/conda/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/conda/lib/python3.8/site-packages/ray/_private/runtime_env/working_dir.py", line 144, in _create
    local_dir = download_and_unpack_package(
  File "/opt/conda/lib/python3.8/site-packages/ray/_private/runtime_env/packaging.py", line 480, in download_and_unpack_package
    raise IOError(f"Failed to fetch URI {pkg_uri} from GCS.")
OSError: Failed to fetch URI gcs://_ray_pkg_e943a30699cdd90c.zip from GCS.


Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
```

**Comment by pouyajafarian** [comment]
We are also seeing this issue:

(raylet, ip=10.128.0.60) [2022-06-06 14:29:29,738 E 364 364] (raylet) agent manager.cc:190: Failed to create runtime env: {"pythonRuntimeEnv": ("pyModules": ["gcs://_ray_pkg_d33eda2267ad8a3b.zip"
, "gcs://_ray_pkg_7abcbseesec3d7c.zip"I}, "uris": ("pyModulesuris": ("ges://_ray_pkg d330da2267adBa3b.zip", "gcs://_ray_pkg_7ab7cbaee:ec3d7c.zip"])], error message: Failed to fetch URI
gcs://_ray pkg d330da2267ad8a3b.zip from GCS
(raylet,
-10.128.0.60) [2022-06-06 14:29:29,738 E 364 364] (raylet) worker pool.cc:623: [Eagerly] Couldn't create a runtime e
nvironment for job eze00000.
(raylet,
ip-10.128.0.60) [2022-06-06 14:29:29,739 E 364 364] (raylet) agent manager.cc:190: Failed to create runtime env: ("pyt
honRuntimeEnv": ("pyModules": ["gcs:// ray pkg d330da2267ad8a3b.zip"
"gcs:// ray pkg 7ab7cb8008ec3d7c.zip"I), "uris": ("pyModulesuris": ["gcs://_ray_pkg_ d330da2267ad8a3b.zip", "gcs://_ray_pkg_7ab7cb8008ec3d7c.zip"]}}, error message: Failed to fetch URI
gcs://_ ray pkg d33eda2267ad8a3b.zip from GCS.
(raylet,
ip-10.128.0.60) [2022-06-06 14:29:29,739 E 364 364] (raylet) worker pool.cc:623: [Eagerly] Couldn't create a runtime e
1 12022-00-00 16:29129.739 304 564 rayler workernvironment for iob 01000000.

For context we are using ray 1.11. Happy to provide more information.

**Comment by architkulkarni** [comment]
We believe https://github.com/ray-project/ray/commit/a67c8a07391ce39d8c656e11b27913150ebe4793 should mitigate this issue.  If you would like to test it out, you can use the [Ray nightly wheels](https://docs.ray.io/en/latest/ray-overview/installation.html#daily-releases-nightlies) on the local machine and all cluster machines.  You'll need to make sure the Ray commit is the same on the local machines and on the cluster; you can check this with `ray.__commit__` in Python.  Please let us know if the issue persists on the nightly wheels!

**Comment by xychu** [comment]
> I too have experienced this, I am using ray 1.12.0. Not sure how related our root causes are however.
> 
> I got it after deleting the nodes using AWS console, the nodes I deleted were the ones hosting my two worker pods.
> 
> I was able to restore the functionality by deleting the head pod.
> 

After a quick look at https://github.com/ray-project/ray/commit/a67c8a07391ce39d8c656e11b27913150ebe4793 , @architkulkarni as @Nintorac mentioned, this happens when deleting the nodes, any clues why that might cause runtime_env got deleted/GCed?

**Comment by architkulkarni** [comment]
Not sure...  The GCS lives on the head node, and the only way the files could be deleted from the head node is if the reference count for the runtime env URI went to zero.  Since the reference count is only decremented when a [detached actor is destroyed](https://github.com/ray-project/ray/blob/af895b3676f759f9f576a2aae8c6887e61ea72f1/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L727) or when a [job exits](https://github.com/ray-project/ray/blob/e9755d87a641b02daf82604ef23ce09647158cd4/src/ray/gcs/gcs_server/gcs_job_manager.cc#L86), I don't see how deleting a node would decrement the reference count.  It might have just been a coincidence that a job exited around the time the node was deleted. 

If the issue is reproducible by deleting worker nodes, that would be very interesting.

**Comment by DmitriGekhtman** [comment]
We should try to figure out a minimal reproduction.

**Comment by hora-anyscale** [comment]
@architkulkarni - per triage discussion, could you please try and repro this on Ray 2.0?

**Comment by architkulkarni** [comment]
We haven't been able to reproduce it but believe we understand the root cause, and https://github.com/ray-project/ray/pull/24719 should fix it.  We haven't had any reports of the issue since then, so I'll close this issue for now, but we should reopen it if we hear any other reports.
